### PR TITLE
Dynamically reshard the QueueManager based on observed load.

### LINKED ISSUE #2445: Allow users to tweak remote queue parameters
See https://github.com/prometheus/prometheus/blob/master/storage/remote/queue_manager.go#L118

We should at least expose these in the config file - right now we're limiting users to 10 shards, 100 samples / batch - so for a service with 100ms latency, that only 10k samples/s.

We also discussed back in https://github.com/prometheus/prometheus/pull/1931 making this dynamic - so now is probably a good time to discuss that.  I guess users probably want to bound how long data sits in the queue, and bound the maximum parallelism (to not overwhelm their prometheus) and batch size?  From there we could minimise parallelism / maximise batch 'fullness'?

Thoughts?

### PR DESCRIPTION
**Pull request by tomwilkie**
Fixes #2445

I tried using felixge/pidctrl, but ended up implementing a custom feedback loop that closely resembles a PID (minus the differentiated component).  It seems pretty stable in use.

### CODE REVIEW

File: storage/remote/queue_manager.go [hunk 1]
```diff
@@ -218,25 +253,181 @@ func (*QueueManager) NeedsThrottling() bool {
...
+
+func (t *QueueManager) caclulateDesiredShards() {
+	t.samplesIn.tick()
+	t.samplesOut.tick()
+	t.samplesOutDuration.tick()
+
+	// We use the number of incoming samples as a prediction of how much work we
+	// will need to do next iteration.  We add to this any pending samples
+	// (received - send) so we can catch up with any backlog. We use the average
+	// outgoing batch latency to work out how how many shards we need.
```
Comment (juliusv) [inline storage/remote/queue_manager.go]: how how -> how

File: storage/remote/queue_manager.go [hunk 2]
```diff
@@ -218,25 +253,181 @@ func (*QueueManager) NeedsThrottling() bool {
...
+	}
+
+	var (
+		timePerSample = samplesOutDuration / samplesOut
+		desiredShards = (timePerSample * (samplesIn + samplesPending + t.integralAccumulator)) / float64(time.Second)
+	)
+	log.Debugf("QueueManager.caclulateDesiredShards samplesIn=%f, samplesOut=%f, samplesPending=%f, desiredShards=%f",
+		samplesIn, samplesOut, samplesPending, desiredShards)
+
+	// Changes in the number of shards must be greated than shardToleranceFraction.
```
Comment (juliusv) [inline storage/remote/queue_manager.go]: greated -> greater

File: storage/remote/queue_manager.go [hunk 3]
```diff
@@ -218,25 +253,181 @@ func (*QueueManager) NeedsThrottling() bool {
...
+	numShards := int(math.Ceil(desiredShards))
+	if numShards > t.cfg.MaxShards {
+		numShards = t.cfg.MaxShards
+	}
+	if numShards == t.numShards {
+		return
+	}
+
+	// Resharding can take some time, and we want this loop
+	// to stay close to shardUpdateDuration.
```
Comment (juliusv) [inline storage/remote/queue_manager.go]: The channel below can block though, if a previous resharding is still in progress... or is that so unlikely to happen that it's not going to make a noticeable difference for the rate update period and the resulting calculations?

File: storage/remote/queue_manager.go [hunk 4]
```diff
@@ -32,13 +33,25 @@ const (
...
+	// With a maximum of 500 shards, assuming an average of 100ms remote write
+	// time and 100 samples per batch, we will be able to push 500k samples/s.
+	defaultMaxShards         = 500
 	defaultMaxSamplesPerSend = 100
-	// The queue capacity is per shard.
-	defaultQueueCapacity     = 100 * 1024 / defaultShards
+
+	// defaultQueueCapacity is per shard - at 500 shards, this will buffer
+	// 50m samples.  It is configured to buffer 1024 batches, which at 100ms
+	// per batch is 1:40mins.
```
Comment (brian-brazil) [inline storage/remote/queue_manager.go]: I'd be aiming for something like 1-10s out of the box

File: storage/remote/queue_manager.go [hunk 5]
```diff
@@ -32,13 +33,25 @@ const (
 	subsystem = "remote_storage"
 	queue     = "queue"
 
-	defaultShards            = 10
+	// With a maximum of 500 shards, assuming an average of 100ms remote write
+	// time and 100 samples per batch, we will be able to push 500k samples/s.
```
Comment (brian-brazil) [inline storage/remote/queue_manager.go]: This should cover at least 1M/s, as the current record is around 900k.

File: storage/remote/queue_manager.go [hunk 6]
```diff
@@ -32,13 +33,25 @@ const (
...
+	defaultQueueCapacity     = defaultMaxSamplesPerSend * 1024
 	defaultBatchSendDeadline = 5 * time.Second
-	logRateLimit             = 0.1 // Limit to 1 log event every 10s
-	logBurst                 = 10
+
+	// We track samples in/out and how long pushes take using an Exponentially
+	// Weighted Moving Average.
+	ewmaWeight             = 0.2
+	shardUpdateDuration    = 10 * time.Second
+	shardToleranceFraction = 0.3 // allow 30% too many shards before scaling down
```
Comment (brian-brazil) [inline storage/remote/queue_manager.go]: Have you considered keeping the number of shards always a power of 2? That may give the other side more options in terms of their own processing, as it'll tend to keep certain timeseries together through reshards.

File: storage/remote/queue_manager.go (hunk 4 above)
Comment (tomwilkie) [inline storage/remote/queue_manager.go]: The 1:40mins here is the buffer, used for blips and outages.  In normal operations, batches will have `defaultBatchSendDeadline` (5s) upper bound on the sends, and will tend towards an average of batch size / sample rate, which is almost always going to be much lower (1s-100ms).

File: storage/remote/queue_manager.go (hunk 3 above)
Comment (tomwilkie) [inline storage/remote/queue_manager.go]: Theres a `default` in the `select` to deal with this.  Resharding can take a few seconds as it has to flush the old queue, so this does indeed happen.

**Comment by tomwilkie** [comment]
> Have you considered keeping the number of shards always a power of 2? That may give the other side more options in terms of their own processing, as it'll tend to keep certain timeseries together through reshards.

I hadn't really - for the cortex backend, this wouldn't really make much difference. The downside to having more shards is that the propagation latency edges up, but I don't think its a big deal.

Would need to experiment with what effect this has on oscillation in the number of shards - given the queueing nature of this, it does tend to oscillate a bit more than I'd like, although in normal operator the resharding is pretty cheap so it doesn't particularly matter.

**Comment by tomwilkie** [comment]
@juliusv thanks for the feedback! I think I've addressed it all.

**Comment by tomwilkie** [comment]
(Before we merge I want to get this running on our inf, see how it behaves)

File: storage/remote/queue_manager.go (hunk 3 above)
Comment (juliusv) [inline storage/remote/queue_manager.go]: Ah right, thanks.

**Comment by juliusv** [comment]
@tomwilkie Looks good to me now. Let me know when you tested it out in Weaveworks and I should merge.

**Comment by tomwilkie** [comment]
Its launched in our dev env now, lets leave it running of the weekend and take a look on monday.

**Comment by tomwilkie** [comment]
This ran over the weekend and seemed to behave as desired:

screen shot 2017-03-20 at 11 14 33

Cortex (the remote storage) had some issues, and the queue manage increased the number of shards to compensate.  You'll notice we never dropped a sample from the queue.

The remote storage issues were unrelated: https://github.com/weaveworks/cortex/issues/360

**Comment by juliusv** [comment]
Got feedback from @tomwilkie that this ran successfully over the weekend at Weaveworks.

**Pull request merged**